source/server/listener_manager_impl.cc

```diff
@@ -473,21 +476,29 @@ void ListenerManagerImpl::drainListener(ListenerImplPtr&& listener) {
   // Tell all workers to stop accepting new connections on this listener.
   draining_it->listener_->debugLog("draining listener");
   const uint64_t listener_tag = draining_it->listener_->listenerTag();
-  stopListener(*draining_it->listener_, [this, listener_tag]() {
-    for (auto& listener : draining_listeners_) {
-      if (listener.listener_->listenerTag() == listener_tag) {
-        // Handle the edge case when new listener is added for the same address as the drained
-        // one. In this case the socket is shared between both listeners so one should avoid
-        // closing it.
-        const auto& socket = listener.listener_->getSocket();
-        if (!hasListenerWithSocket(active_listeners_, socket) &&
-            !hasListenerWithSocket(warming_listeners_, socket)) {
-          // Close the socket iff it is not used anymore.
-          listener.listener_->socket().close();
+  stopListener(
+      *draining_it->listener_,
+      [this,
+       share_socket = draining_it->listener_->listenSocketFactory().sharedSocket().has_value(),
```

Comment:
It's a little odd to me that are are capturing share_socket (nit should be shared_socket) here, only to return on it in the below lambda. Can you just determine this from within the lambda since you look up the listener anyway? More comments would be better also. Same below.

Reply:
The lambda is capturing a bool rather than the socket. The early return can avoid matching listener_tag if the socket is not shared. 

Added more comment about the share_socket.